### ISSUE TITLE
throw an error if you use `unwrap` without checking first if you have a `none` or `some`

### ISSUE THREAD

**Issue by ryanguill** [iwburns/nullshield#44]
I'm not sure this is a good idea or not yet, but what if we keep track per instance if a given Option has been "checked", and if you attempt to `unwrap` before checking we throw an error, even if it is a `some`.

If you are working with ts, this probably isnt as much of a concern - but if you are working from javascript directly, theres nothing to prevent you from misunderstanding the api and not checking first.  If you just happen to only test with values that produce `some`s, you might never realize you forgot a check to make sure thats safe first.

Imagine a newcomer writing code like this:

```
service.methodThatReturnsAPromiseOfAnOption()
   .then(o => o.unwrap())
   .then(doSomethingCool);
```

They're new to `Option`s, they read enough of the api to know that `unwrap` is how they get the value they care about out and they move along.  

The error could say something like 
> You attempted to unwrap an `Option` without first checking to make sure it has a value! see this url for more information, but consider using `unwrapOr` or using `.isSome()` or `.isNone()` to test if it is safe to `unwrap` first!

I'm not certain if there would be edge cases where they _could_ know its safe but we cant know that they know - might be something around where we create new instances in some operations, but would need to think about that more - most of those operations could return null and so the point of creating a new instance is to make that a safe operation - meaning if you did check before (on an earlier instance) that doesn't mean you are still safe now.

Also not certain if there should be a way to say "leave me alone, I know what im doing" - or what it would look like if there were.

**Comment by iwburns** [comment]
Letting this sit in my mind for a bit as I'm not sure how I feel about it yet.  I think I like where you're going, but I'm not sure about those edge-cases you mentioned.  Will get back to you on this soon.

**Comment by iwburns** [comment]
I like the spirit of this, but I'm not sure I think this is the right solution to the problem.  Some drawbacks I can see are:

1. extra processing and storage internally (maybe it's not enough to matter, but still something to consider)
2. slightly more complexity to the logic in the library (again maybe not enough to matter, but still)
3. you already basically mentioned this, but what if I'm prototyping something and I'm fine with just using `unwrap()` temporarily (assuming I know what I'm doing).
4. there are definitely edge-cases that we probably won't be able to detect (or even think of).  And as long as these exist, what can we really do?  We can't throw an error because if there _is_ an edge-case, then people who run into that can't use the library at all, right? (unless I'm missing something).

However, I do like the error message you wrote up and I don't see why we couldn't just use that (maybe slightly tweaked) for the normal error message when you unwrap a `None`.  I think that would help a great deal when someone _does_ get to the point of unwrapping what they thought was a `Some`.

The root of the problem here though is preventing that scenario before it even happens, but I'm not sure if that's really possible in all scenarios.  IMHO, we just need to make this as easy as possible for people who are new to Optional types, and I think that is best solved with good documentation.

Italics / bold in the docs for `unwrap` specifically would probably help.  And it would also be really good to have a section front and center both in the readme and in the docs called "New to Option-types?" or something like that.  It would also probably be nice to have a section for people who aren't new to Optional types which just gives the high-level overview so they know which set of terminology this library uses (Maybe/Option, Just/Some, Nothing/None, get/unwrap, etc.).

**Comment by ryanguill** [comment]
I'm still not convinced this is a good idea either, but let me argue for it a little longer.

re: 1 and 2, those are not good reasons imo.  The impact to storage, processing and logic is tiny.  We are talking about a boolean field intialized to false when a new Option is instantiated and flipped to true when you call certain methods - with a single if when you call `unwrap` and only then.

> what if I'm prototyping something and I'm fine with just using unwrap() temporarily (assuming I know what I'm doing)

You don't know what you're doing... unsafe code is unsafe code.  There is no way to call unwrap safely without first checking that it contains a `some`.  Even in prototype code, its just wrong.  In a type safe language you would be prevented from doing it.  And I mean thats kinda the point of using this this pattern in the first place - you need more safety in your life.  If you want to play it fast and loose, why would you use an `Option`?

> there are definitely edge-cases that we probably won't be able to detect (or even think of). And as long as these exist, what can we really do? We can't throw an error because if there is an edge-case, then people who run into that can't use the library at all, right?

I haven't been able to come up with a single one.  Even a hypothetical.  The only time I can conceive of that it would happen is if there were other methods you could call that would let you know that it is indeed safe to call `unwrap` besides `isSome` and `isNone`.  I don't think there are, but if we find that is the case, we add the setting of the flag to that method and we're done.  There shouldn't ever be a case you couldn't fall back to using `isSome` and `isNone` though.

>The root of the problem here though is preventing that scenario before it even happens, but I'm not sure if that's really possible in all scenarios. IMHO, we just need to make this as easy as possible for people who are new to Optional types, and I think that is best achieved with good documentation. 

Even if you read the documentation and think you understand it, you could still accidentally make this mistake.  And in your testing you never hit a case where the Option is `none` and so you dont notice.  And then you blow up in production - so what good was this library at that point?  Again, in other languages you have the luxury of the compiler being able to tell you if you do this stupid thing - we have to do it at runtime.

The documentation improvements would still be nice though - if we don't end up doing anything but that then I think this has been a worthwhile exercise.  But I'll assert that most people don't read the documentation and they write some code and feel around till they get something working and they ship it.

**Comment by iwburns** [comment]
>re: 1 and 2, those are not good reasons imo.[...]

Fair points here, just wanted to bring it up.  I think it's slightly more complicated than that, but more on that later.

>You don't know what you're doing... unsafe code is unsafe code[...]

I get this argument (and I kindof agree with it), but I feel like it's a bit too idealistic to say that without any caveats at all.  Generally speaking, I don't want people to use a given library improperly nor do I want them to build bad habits; but I also don't want to try to control how people write their own programs _too_ much.  I think it's fine (and very much a good idea) to design an API such that we encourage "proper" usage, but I also wouldn't want to make it more difficult than necessary to use the API just to be "pure" about the approach.

> In a type safe language you would be prevented from doing it.[...]

I'm not sure this is entirely true.  (safe) Rust is type-safe (iirc), and you can absolutely call `unwrap` on a `None` value.  It's just that the `unwrap` call will result in a `panic` which will dump the stack and halt the program.  The supposed return type of `unwrap` is still `T`, so the types don't mismatch and there's no compile-time issue here.  Regardless though:

>If you want to play it fast and loose, why would you use an Option?

I totally agree with the sentiment here.  I think the only issue is, if you are using a library that provides an API that already uses Options, you don't have a choice.

>I haven't been able to come up with a single one.[ ... ]if there were other methods you could call that would let you know that it is indeed safe to call `unwrap` besides `isSome` and `isNone`

So, just to list out the items we'd need to modify:
* filter
* flatMap
* forEach
* isNone
* isSome
* map
* match
* orElse
* unwrapOr
* unwrapOrElse

Basically, anything that takes a function as an argument (we don't know what side-effects those functions can cause), `isNone` / `isSome`, and `unwrapOr` because you could then check the return value and then optionally call `unwrap`.  

However, let's say that we add that logic to all of those functions, what happens when someone writes this:
```
// initialized "hasChecked" to false;
const nope = OptionT.none();

// set "hasChecked" to true because they may have saved some information from within that lambda
const nada = nope.map(val => val * 2);

// now they can call unwrap without getting the special error message
const originalVal = nope.unwrap();
```
With that in mind, I'm not sure how we could consistently prevent the consumer from running into this issue.  We might be able to in some cases, but there are definitely cases where we can't know whether or not they know what the state of the OptionT is in.  We know they _could_ know, but not whether or not they _do_ know.

**Comment by ryanguill** [comment]
I don't think we are on the same page.

None of those methods besides isNone or isSome will tell you if you have a some or none.  Lets take `map` for instance.

https://repl.it/repls/WigglySelfassuredGnuassembler
```
let a = OptionT.some(1);
let b = OptionT.some(1).map(x => x > 2 ? x : null); 
console.log(b);// None()
```
You could have known you had a `some` - but then once you `map` you dont know that any longer.  Remember that it _has_ to be this way because all of these methods return new `Option`s, not the same one.  Even if we were to keep track of `hasChecked ` with these methods it wouldn't matter because we would then just create a new Option for the resulting value.

Saving information from inside of lambdas is not something we should support.  Even if you could do it reliably, thats not how it should be used.  Remember that in the example you gave the lambda wasn't even called at all, because it was a `none`.

**Comment by iwburns** [comment]
So, you're correct, we weren't on the same page.  In your example above, I agree that the only way you can tell what `b` is would be to call `isSome` or `isNone`.

However, I think my previous point wasn't quite clear and it may still be of some concern.  Consider this slightly tweaked example:

https://repl.it/repls/RobustElementaryPentagon
```
let a = OptionT.some(5);
let b = a.map(x => x > 2 ? x : null); 

if (b.isSome()) {
  console.log(a.unwrap());
}
```
or even more contrived:
```
let canUseA = false;

let a = OptionT.some(1);
let b = a.map(x => {
  canUseA = true;
  return x * 2;
});

if (canUseA) {
  console.log(a.unwrap());
}
```
In those scenarios, we'd be preventing them from unwrapping `a` even though they know that it's safe to do so.

**Comment by ryanguill** [comment]
So your second example is exactly what im saying we shouldn't support.  We have a way for them to determine if they can use `a` or not, its `isSome` or `isNone` - imo we shouldn't worry about that use case at all.

I can _kinda_ see where you are coming from on your first example - I think just more simply stated youre saying that because they are the ones using `.some()` to create the option then they know its a `Some` - but I don't think this is really a representative case. I think 95% of the time or more, the consumer, the one doing the `unwrap`ing is not going to be the one that created the Option in the first place - some other code has done that and returned the Option to you.

Maybe a middle ground is to make another method besides `unwrap` that allows you to declare that you know what you're doing.  `unsafeUnwrap()` or `unwrapOrThrow()` or something like that. Basically like an assertion, I know this is safe even though you don't think I should be able to.  But at the end of the day, is it really that bad to force them to use `isSome/isNone` or `unwrapOr`? The spirit of the library is safety - my goal is just to guide them in how to use it safely.

**Comment by iwburns** [comment]
> So your second example is exactly what im saying we shouldn't support.

Understood, and I thought about that as I was writing it, but wasn't 100% sure so I figured I'd included it anyway.  I have mixed feelings on this.  Part of me wants to be "pure" about this and guarantee that a feature like this can't get in the way of someone doing something valid even if what they're doing is strange or not quite the recommended way to do things.  However, I do understand that - as you said (slightly tweaked) - the whole point of this library _is_ to get in people's way and make them do things differently.

>[...]youre saying that because they are the ones using .some() to create the option[...]

Actually, I see how you got there, but that wasn't really my point.  I was trying to show that in certain cases they could check one option and therefore know information about another option.  For example:
https://repl.it/@iwburns/FluidTurquoiseValue
```
function getOption() {
  if (Math.random() > 0.5) {
    return OptionT.some(1);
  }
  return OptionT.none();
}

let a = getOption();
let b = a.map(x => x * 2);

if (b.isSome()) { // check b, but get info about a
  console.log(a.unwrap());
} else {
  console.log('a is a none');
}
```
In this scenario, it is guaranteed that if `b.isSome()` is `true`, `a.isSome()` will be too.  In fact, I _think_ this is true regardless of the lambda passed to `map` (as long as the lambda doesn't reassign `a`). 

**Side Note (just a fun observation)**: The opposite isn't necessarily true, in other words if `b.isSome()` is `false` we don't know that `a.isSome()` would also be `false`.

I do think that the above isn't likely to be very common though.  And I agree, I think a good middle ground here would be to have a separate function that behaves as `unwrap()` currently does and the change the way `unwrap()` behaves as you described originally.  I'm not against `unsafeUnwrap()` and `unwrapOrThrow()`, but just to toss another name out there; what do you think of `forceUnwrap()`?

**Comment by iwburns** [comment]
@ryanguill I've been busy lately and haven't looked at this in a bit. Do you have any more thoughts on this approach?

**Comment by iwburns** [comment]
On mobile, so my formatting may be poor.

Your filter example is a good one. The only other thing I can think of would be an if condition with an `and`/`or` chain with an `isSome` at the end and then a bunch of unwraps in the if-body.

Because of the way that `and` and `or` work, the `isSome` would only affect the one that it happened to get called on and we wouldn't know exactly which one without first knowing whether or not they're  a some or a none.

**Comment by iwburns** [comment]
updating this with a code example now that I'm not on mobile:
```
let a = OptionT.some(1);
let b = OptionT.some(2);
let c = OptionT.some(3);

if (a.and(b).and(c).isSome()) {
  // here we know that a, b, and c are all okay to unwrap
  // but only c would be marked as one that we've checked.
  doStuff(a.unwrap(), b.unwrap(), c.unwrap());
}
```
The more I think about it, `or` _shouldn't_ be an issue here.  Logically it wouldn't make much sense to do:
```
if (a.or(b).or(c).isSome()) {
  // here we only know that one of them is okay to unwrap.
  // so blindly unwrapping all of them here is a bad idea and could very likely throw an error already
  doStuff(a.unwrap(), b.unwrap(), c.unwrap());
}
```

With that said though, I'm not sure how often the first example here would even come up.  I've actually never written code like that before while using `Option`-types, but that doesn't mean it never _could_ be the right thing to do.

**Comment by iwburns** [comment]
>I _think_ what we can do is [...]

I believe you are correct.  I'm not seeing how you could get into a weird scenario with this, do you mind giving an example?  Even so, I agree that it probably won't be too hard for someone to figure out what's going on.

So, summarizing the current path forward (assuming we implement this), we need to modify `isSome`, `isNone`, and `and` to all set the internal "checked" flag to `true`.  In addition, we need to modify `unwrap` to throw a special message if the internal "checked" flag is `false` on both `Some` and `None` types.  Otherwise, those methods should follow their normal behavior.

Does that all sound right?

**Comment by ryanguill** [comment]
the scenario is something like this:

```
let a = OptionT.some(1);
let b = OptionT.none();

let c = a.and(b);
doStuff(a.unwrap);// we have set the flag on `a` because you called `and` so we let this `unwrap` happen.
```

I think your summary is correct, but we also need to implement `forceUnwrap` which also throws a console warning that it isnt intended for production use, discouraged, etc.

**Comment by iwburns** [comment]
Ah, yep that would do it.  And I agree, I don't see any problem with allowing the unwrap call there.  Worst case scenario is that `a` is actually a `none` and they get the "normal" called-unwrap-on-a-none error message.

And good call about `forceUnwrap`; somehow forgot about that one.  My main use-case example for wanting a way to `forceUnwrap` something is for writing tests or something similar where if the unwrap fails, the test (or whatever) should probably fail as well.  I think a `console.warn` is probably fine in that case, but I wanted to run it by you just in case you had any thoughts on that.

**Comment by ryanguill** [comment]
not sure I understand what youre after, but you can capture thrown errors in your unit tests and you can also mock the console object.

**Comment by iwburns** [comment]
I'm just making a case for when I think using `forceUnwrap` would be acceptable.  Let's say you're writing a test for the following function:
```
function safeParseInt(value) {
  const parsed = Number.parseInt(value, 10);
  if (Number.isNaN(parsed) {
    return OptionT.none();
  }
  return OptionT.some(parsed);
}
```
You might then write the following tests:
```
let parsed = safeParseInt('1');
expect(parsed.forceUnwrap()).to.equal(1);

parsed = safeParseInt('-1');
expect(parsed.forceUnwrap()).to.equal(-1);
```
You'll get some warnings from this, but I think this is a perfectly reasonable test to write.  If the unwrap fails, it means your function isn't doing what it's supposed to.

One could argue that you should instead write:
```
let parsed = safeParseInt('1');
parsed.match({
  some: val => { expect(val).to.equal(1); },
  none: () => { throw new Error('safeParseInt gave a None when a Some(1) was expected.'); },
});
```
But I think that's not really that much more useful, especially for those more simple unit tests.

With all of that said, do you think it's reasonable to expect console warnings from tests that use that strategy?  I'm personally a bit torn.

**Comment by ryanguill** [comment]
I think you should instead write
```
let parsed = safeParseInt('1');
expect(parsed.unwrapOr(NaN)).to.equal(1);
```

If you used `unwrap` or `forcedUnwrap` and it is a `none`, meaning your test fails, you will get an error - which depending on the test suite might bomb the test instead of failing the assertion.  Using `unwrapOr` allows you to easily provide some other value than the expected to check against.

My personal opinion is `forceUnwrap` should be used for quick prototyping (not including tests) or as an escape hatch, and if you need the escape hatch we should be encouraging you to file an issue here so we can fix the underlying issue.

**Comment by iwburns** [comment]
>[...] which depending on the test suite might bomb the test instead of failing the assertion [...]

Fair point, I didn't think about that.  And I agree that `unwrapOr` is a very good candidate to use in this example (and probably most others).  Thinking more on it, the only situation I can think of where `unwrapOr` isn't what you want is when you're dealing with objects or arrays.  And in those cases, `forceUnwrap` wouldn't be what you would want either; instead you'd probably need to `filter` or `match` (or something similar) to get your desired information.

>My personal opinion is [...]

I think I agree here.

I don't really have anything to add to the discussion here at this point.  How do you feel about this feature now that we've (I think fully?) talked it out?

**Comment by ryanguill** [comment]
I still think it would be worthwhile.  The only thing I think we are adding to the api is the `forceUnwrap`, everything else we could walk back from if we find that its not a good idea for some reason.  I think its a good way to help make sure people aren't using it incorrectly.

**Comment by iwburns** [comment]
I agree.  If we had to walk it back in the future, worst case scenario is that `forceUnwrap` and `unwrap` end up doing the same thing.  That's obv. not ideal and I kindof doubt that will happen, but even if it does it's not _that_ big of a deal.

Also, if we did end up walking this back and want to remove `forceUnwrap`, I'd say that's a pretty minor breaking change.  It would of course still warrant a major version bump, but the migration guide for that change would be pretty straightforward.

I'm game to move this forward past the discussion phase if you are.  I don't really care who works on it.  I should have some free time over the next few days to implement this, but if you have time feel free to move forward on it as well.  Just let me know so we don't double work it; I'll do the same whenever I sit down to work on it.
